File: Sem3/Advanced Methods of Programming/MyInterpreter - JavaFX - Final Project/project/src/domain/statement/ControlStatements/IfStatement.java
package domain.statement.ControlStatements;


import domain.ProgramState;
import domain.adt.MyDictionaryInterface;
import domain.adt.MyStackInterface;
import domain.expression.Expression;
import domain.expression.ExpressionException;
import domain.statement.StatementException;
import domain.statement.StatementInterface;


public class IfStatement implements StatementInterface
{
    private Expression expression;
    private StatementInterface thenStatement;
    private StatementInterface elseStatement;


    public IfStatement(Expression expression, StatementInterface thenStatement,
                       StatementInterface elseStatement)
    {
        this.expression = expression;
        this.thenStatement = thenStatement;
        this.elseStatement = elseStatement;
    }


    @Override
    public String toString()
    {
        return "IF(" + expression.toString() + ")THEN(" + thenStatement.toString()
                + ")ELSE(" + elseStatement.toString() + ")";
    }


    @Override
    public ProgramState execute(ProgramState programState) throws StatementException
    {
        MyStackInterface<StatementInterface> stack = programState.getExeStack();
        MyDictionaryInterface<String, Integer> symbolTable = programState.getSymbolTable();
        MyDictionaryInterface<Integer, Integer> heapTable = programState.getHeapTable();
        try
        {
            if (expression.evaluate(symbolTable, heapTable) != 0)
                stack.push(thenStatement);
            else
                stack.push(elseStatement);
        }
        catch (ExpressionException exception)
        {
            throw new StatementException(exception.getMessage() + " If statement.");
        }
        return null;
    }


    public Expression getExpression()
    {
        return expression;
    }
}

File: Sem3/Advanced Methods of Programming/MyInterpreter - CSharp Console Application/MyInterpreter CSharp/domain/statement/IfStatement.cs
using System;
using System.Collections;
using MyInterpreter_CSharp.domain.adt;
using MyInterpreter_CSharp.domain.expression;

namespace MyInterpreter_CSharp.domain.statement
{
    public class IfStatement : IStatement
    {
        private Expression _expression;
        private IStatement _thenStatement;
        private IStatement _elseStatement;

        public IfStatement(Expression expression, IStatement thenStatement, IStatement elseStatement)
        {
            _expression = expression;
            _thenStatement = thenStatement;
            _elseStatement = elseStatement;
        }

        public ProgramState Execute(ProgramState programState)
        {
            IMyStack<IStatement> stack = programState.ExeStack;
            IMyDictionary<string, int> symbolTable = programState.SymbolTable;
            try
            {
                if (_expression.Eval(symbolTable) != 0)
                    stack.Push(_thenStatement);
                else
                    stack.Push(_elseStatement);
            }
            catch (ExpressionException exception)
            {
                throw new StatementException(exception.Message + " " + _expression);
            }

            return null;
        }

        public override string ToString()
        {
            return "IF(" + _expression + ")THEN(" + _thenStatement + ")ELSE(" + _elseStatement + ")";
        }
    }
}